src/Type/Definition/ObjectType.php

```diff
@@ -117,6 +118,11 @@ public static function assertObjectType($type)
      */
     public function getField($name)
     {
+        $typeNameDef = Introspection::typeNameMetaFieldDef();
+        if ($name === $typeNameDef->name) {
+            return $typeNameDef;
+        }
+
```

Comment:
Somehow I managed to miss that you added changes to `ObjectType`. This is not quite correct because introspection is a completely separate process. `ObjectType` shouldn't know anything about it. Instead, we should literally check for `__typename` field in the `QueryPlan` and `getFieldSelection` method.

Reply:
Ok, no problem! I have updated the QueryPlan and the getFieldSelection method already provides the expected behavior.
I don't know the code in depth, maybe I've been cheeky to write a PR, tell me everything you don't see correctly or feel free to rewrite it.